src/shamir.rs

```diff
@@ -0,0 +1,419 @@
+    #[error("Wrong or insecure secret sharing scheme. (expected {k} >= {n} > 1)")]
+    BadSharingScheme { k: u8, n: u8 },
+    #[error("The degree of polynomial provided to reconstruct secret does not match number of points provided. \
+     Polynomial degree: {polynomial_degree}, number of points {points_count}")]
+    BadPolynomial {
+        polynomial_degree: u8,
+        points_count: u8,
+    },
+    #[error("The degree of Lagrange polynomial {polynomial_degree} is greater than the field size {field_size}")]
+    FieldSizeError {
+        field_size: u8,
+        polynomial_degree: u8,
+    },
+    #[error("Prime field element inversion failed: {v:?}")]
+    InvertError { v: Box<[u8]> },
+}
+
+/// Represents a single share: f(x) point. The index of it inside the share slice is used to represent
+/// the "x" coordinate of this share.
+#[derive(Clone)]
+pub struct Share<F> {
+    y: F,
+}
+
+impl From<Share<Fp31>> for Fp31 {
+    fn from(share: Share<Fp31>) -> Self {
+        share.y
+    }
+}
+
+impl SecretSharing {
+    /// Constructs a new instance, returning `ShamirError` if input values do not allow
+    /// to construct a valid Shamir secret sharing scheme.
+    ///
+    /// # Errors
+    /// Rejects the following inputs:
+    /// * k == 1
+    /// * k > n
+    pub fn new(k: NonZeroU8, n: NonZeroU8) -> Result<Self, Error> {
+        if k.get() == 1 || k > n {
+            Err(Error::BadSharingScheme {
+                k: k.get(),
+                n: n.get(),
+            })
+        } else {
+            Ok(Self {
+                k: k.get(),
+                n: n.get(),
+            })
+        }
+    }
+
+    /// Minimum number of shares required to reconstruct a secret
+    #[must_use]
+    pub fn threshold(&self) -> u8 {
+        self.k
+    }
+
+    /// Consumes and splits a given secret into `n` shares. only `k` <= `n` shares is required
+    /// to reconstruct it later
+    pub fn split<R: RngCore, F: Field>(&self, secret: F, rng: R) -> Vec<Share<F>> {
+        // generate polynomial of k-1 degree
+        let coefficients = Self::gen_polynomial::<R, F>(rng)
+            .take(usize::from(self.k - 1))
+            .collect::<Vec<_>>();
+
+        let mut shares = Vec::with_capacity(self.n as usize);
+
+        // sample n points by evaluating the polynomial
+        for i in 1..=self.n {
+            let mut y = F::ZERO;
+            let x = F::from(u128::from(i));
+            for c in coefficients.iter().rev() {
+                y += *c;
+                y *= x;
+            }
+
+            y += secret;
+
+            shares.push(Share { y });
+        }
+
+        shares
+    }
+
+    /// Reconstructs a secret from a set of size at least k of shares.
+    ///
+    /// # Errors
+    /// Returns an error if there is no enough shares to reconstruct the secret or
+    /// if an error occurred while evaluating the polynomial
+    pub fn reconstruct<F>(shares: &[Share<F>], lagrange: &LagrangePolynomial<F>) -> Result<F, Error>
```

Comment:
Once again - perhaps add a comment explaining that the order of "shares" matters. It must match the order of the helpers.